cob_obstacle_distance/src/shapes_manager.cpp

```diff
@@ -38,22 +37,51 @@ ShapesManager::~ShapesManager()
 }
 
 
-void ShapesManager::addShape(t_ptr_IMarkerShape s)
+void ShapesManager::addShape(const std::string& id, PtrIMarkerShape_t s)
+{
+    this->shapes_[id] = s;
+}
+
+
+void ShapesManager::removeShape(const std::string& id)
 {
-    this->shapes_.push_back(s);
+    if(this->shapes_.count(id))
+    {
+        PtrIMarkerShape_t s = this->shapes_[id];
+        visualization_msgs::Marker marker = s->getMarker();
+        marker.action = visualization_msgs::Marker::DELETE;
+        this->pub_.publish(marker);
+    }
+
+    this->shapes_.erase(id);
+}
+
+
+bool ShapesManager::getShape(const std::string& id, PtrIMarkerShape_t& s)
+{
+    bool success = false;
+    if(this->shapes_.count(id))
+    {
+        s = this->shapes_[id];
+        success = true;
+    }
+
+    return success;
 }
 
 
 void ShapesManager::draw(bool enforce_draw)
```

Comment:
Drawing is called cyclic. To avoid publishing the same obstacle all the time the isDrawn Property is evaluated and avoids publishing again. enforce_draw==true -> isDrawn will be ignored e.g. for moving obstacles.